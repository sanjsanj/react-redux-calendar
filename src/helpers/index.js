import { reminders } from "../reducers/reminders";

/** Test helpers */
export const styledComponentSelector = className =>
  `[data-class="${className}"]`;

/** Reducer helpers */
export const updateObject = (prevObject, nextObject) => ({
  ...prevObject,
  ...nextObject,
});

export const updateArray = (prevArray, item) => [...prevArray, item];

/** Time helpers */
export const dayNamesArray = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

const _monthNamesArray = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

export const getMonthName = currentMonthIndex =>
  _monthNamesArray[currentMonthIndex];

const _getLeadingDaysArray = daysArray =>
  Array.apply(undefined, Array(daysArray[0].dayNameIndex));

const _getTrailingDaysArray = daysArray => {
  const lastDay = daysArray[daysArray.length - 1];
  const numOfDays = Math.abs(lastDay.dayNameIndex - 6);

  return Array.apply(undefined, Array(numOfDays));
};

const _getDayNameIndex = (year, month, day) =>
  new Date(`${year}-${month + 1}-${day + 1}`).getDay();

export const getDaysForMonth = (year, month) => {
  const daysInMonth = new Date(year, month + 1, 0).getDate();

  const daysArray = Array.apply(undefined, Array(daysInMonth)).map(
    (day, index) => ({
      dayNameIndex: _getDayNameIndex(year, month, index),
      dayOfMonth: index + 1,
    })
  );

  return [
    ..._getLeadingDaysArray(daysArray),
    ...daysArray,
    ..._getTrailingDaysArray(daysArray),
  ];
};

export const ensureTwoDigits = number => `0${number}`.slice(-2);

/** With-data helpers */
export const withReminders = (daysOfTheMonthArray, month, year, reminders) => {
  return daysOfTheMonthArray.map((day, index) => {
    if (!day) return day;

    const startOfDay = new Date(year, month, day.dayOfMonth).getTime();
    const endOfDay = new Date(year, month, day.dayOfMonth + 1).getTime();

    return updateObject(day, {
      reminders: reminders.filter(
        reminder =>
          reminder.dateTime >= startOfDay && reminder.dateTime <= endOfDay
      ),
    });
  });
};

/** Form helpers */
export const formSerialize = formElement => {
  const values = {};
  const inputs = formElement.elements;

  for (let i = 0; i < inputs.length; i++) {
    values[inputs[i].name] = inputs[i].value;
  }
  return values;
};

export const stringTimeToDateTime = time => new Date(time).getTime();
