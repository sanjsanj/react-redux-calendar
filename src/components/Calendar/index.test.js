import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import { getDaysForMonth } from "../../helpers";

import { Calendar } from ".";

describe("Calendar", () => {
  const props = {
    daysOfTheMonth: getDaysForMonth(2019, 3),
    selectedYear: 2019,
    monthName: "April",
  };

  const wrapper = shallow(<Calendar {...props} />);

  it("should display the selected month and year", () => {
    expect(wrapper.text().includes("April 2019")).toEqual(true);
  });

  it("should display correct number of calendar day boxes", () => {
    expect(wrapper.find("CalendarDayBox").length).toEqual(
      props.daysOfTheMonth.length
    );
  });

  it("should match snapshot", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
