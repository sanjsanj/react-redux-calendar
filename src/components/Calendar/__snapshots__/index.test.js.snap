// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar should match snapshot 1`] = `
<Fragment>
  <h2>
    April
     
    2019
  </h2>
  <styled.div>
    <styled.p
      key="0"
    >
      Sun
    </styled.p>
    <styled.p
      key="1"
    >
      Mon
    </styled.p>
    <styled.p
      key="2"
    >
      Tue
    </styled.p>
    <styled.p
      key="3"
    >
      Wed
    </styled.p>
    <styled.p
      key="4"
    >
      Thu
    </styled.p>
    <styled.p
      key="5"
    >
      Fri
    </styled.p>
    <styled.p
      key="6"
    >
      Sat
    </styled.p>
    <CalendarDayBox
      key="0"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 1,
          "dayOfMonth": 1,
        }
      }
      key="1"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 2,
          "dayOfMonth": 2,
        }
      }
      key="2"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 3,
          "dayOfMonth": 3,
        }
      }
      key="3"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 4,
          "dayOfMonth": 4,
        }
      }
      key="4"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 5,
          "dayOfMonth": 5,
        }
      }
      key="5"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 6,
          "dayOfMonth": 6,
        }
      }
      key="6"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 0,
          "dayOfMonth": 7,
        }
      }
      key="7"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 1,
          "dayOfMonth": 8,
        }
      }
      key="8"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 2,
          "dayOfMonth": 9,
        }
      }
      key="9"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 3,
          "dayOfMonth": 10,
        }
      }
      key="10"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 4,
          "dayOfMonth": 11,
        }
      }
      key="11"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 5,
          "dayOfMonth": 12,
        }
      }
      key="12"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 6,
          "dayOfMonth": 13,
        }
      }
      key="13"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 0,
          "dayOfMonth": 14,
        }
      }
      key="14"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 1,
          "dayOfMonth": 15,
        }
      }
      key="15"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 2,
          "dayOfMonth": 16,
        }
      }
      key="16"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 3,
          "dayOfMonth": 17,
        }
      }
      key="17"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 4,
          "dayOfMonth": 18,
        }
      }
      key="18"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 5,
          "dayOfMonth": 19,
        }
      }
      key="19"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 6,
          "dayOfMonth": 20,
        }
      }
      key="20"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 0,
          "dayOfMonth": 21,
        }
      }
      key="21"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 1,
          "dayOfMonth": 22,
        }
      }
      key="22"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 2,
          "dayOfMonth": 23,
        }
      }
      key="23"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 3,
          "dayOfMonth": 24,
        }
      }
      key="24"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 4,
          "dayOfMonth": 25,
        }
      }
      key="25"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 5,
          "dayOfMonth": 26,
        }
      }
      key="26"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 6,
          "dayOfMonth": 27,
        }
      }
      key="27"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 0,
          "dayOfMonth": 28,
        }
      }
      key="28"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 1,
          "dayOfMonth": 29,
        }
      }
      key="29"
    />
    <CalendarDayBox
      day={
        Object {
          "dayNameIndex": 2,
          "dayOfMonth": 30,
        }
      }
      key="30"
    />
    <CalendarDayBox
      key="31"
    />
    <CalendarDayBox
      key="32"
    />
    <CalendarDayBox
      key="33"
    />
    <CalendarDayBox
      key="34"
    />
  </styled.div>
</Fragment>
`;
